pkg/controller/atlasstream/atlasstream_instance_controller.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/statushandler"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+)
+
+const instanceNotFound = "STREAM_TENANT_NOT_FOUND_FOR_NAME"
+
+type AtlasStreamsInstanceReconciler struct {
+	Client                      client.Client
+	Scheme                      *runtime.Scheme
+	GlobalPredicates            []predicate.Predicate
+	EventRecorder               record.EventRecorder
+	AtlasProvider               atlas.Provider
+	Log                         *zap.SugaredLogger
+	ObjectDeletionProtection    bool
+	SubObjectDeletionProtection bool
+}
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasstreaminstances,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasstreaminstances/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasstreaminstances,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasstreaminstances/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasstreamconnections,verbs=get;list
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasstreamconnections,verbs=get;list
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+
+// https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20%20%20subgraph%20cluster_pending%20%7B%0A%20%20%20%20%20%20%20%20skipped%3B%0A%20%20%20%20%20%20%20%20invalid%3B%0A%20%20%20%20%20%20%20%20unsupported%3B%0A%20%20%20%20%20%20%20%20terminated%3B%0A%20%20%20%20%20%20%20%20label%20%3D%20%22pending%22%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20deleted%20%5Blabel%3D%22deleted%5Cnfinalizer%20unset%22%5D%0A%0A%20%20%20%20pending%20-%3E%20pending%20%5Blabel%3D%22skip%5Cninvalidate%5Cnunsupport%5Cnterminate%22%5D%0A%20%20%20%20pending%20-%3E%20ready%20%5Blabel%3D%22create%22%5D%0A%20%20%20%20pending%20-%3E%20deleted%20%5Blabel%3D%22delete%22%5D%0A%20%20%20%20ready%20-%3E%20ready%20%5Blabel%3D%22update%22%5D%0A%20%20%20%20ready%20-%3E%20deleted%20%5Blabel%3D%22delete%22%5D%0A%20%20%20%20ready%20-%3E%20pending%20%5Blabel%3D%22terminate%22%5D%0A%7D%0A
+
+func (r *AtlasStreamsInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.With("atlasstreaminstance", req.NamespacedName)
+	log.Infow("-> Starting AtlasStreamInstance reconciliation")
+
+	akoStreamInstance := akov2.AtlasStreamInstance{}
+	result := customresource.PrepareResource(ctx, r.Client, req, &akoStreamInstance, log)
+	if !result.IsOk() {
+		return result.ReconcileResult(), nil
+	}
+
+	return r.ensureAtlasStreamsInstance(ctx, log, &akoStreamInstance)
+}
+
+// this is the central state dispatcher
```

Comment:
Does our linter not complain if doc strings are not conformant to conventions?

```suggestion
// ensureAtlasStreamsInstance is the central state dispatcher
```